packages/contracts-bedrock/src/safe/DeputyPauseModule.sol

```diff
@@ -0,0 +1,207 @@
+    function foundationSafe() public view returns (Safe foundationSafe_) {
+        foundationSafe_ = FOUNDATION_SAFE;
+    }
+
+    /// @notice Getter function for the SuperchainConfig address.
+    /// @return superchainConfig_ SuperchainConfig address.
+    function superchainConfig() public view returns (ISuperchainConfig superchainConfig_) {
+        superchainConfig_ = SUPERCHAIN_CONFIG;
+    }
+
+    /// @notice Getter function for the Pause message typehash.
+    /// @return pauseMessageTypehash_ Pause message typehash.
+    function pauseMessageTypehash() public pure returns (bytes32 pauseMessageTypehash_) {
+        pauseMessageTypehash_ = PAUSE_MESSAGE_TYPEHASH;
+    }
+
+    /// @notice Getter function for the DeputyAuth message typehash.
+    /// @return deputyAuthMessageTypehash_ DeputyAuth message typehash.
+    function deputyAuthMessageTypehash() public pure returns (bytes32 deputyAuthMessageTypehash_) {
+        deputyAuthMessageTypehash_ = DEPUTY_AUTH_MESSAGE_TYPEHASH;
+    }
+
+    /// @notice Sets the deputy address.
+    /// @param _deputy Deputy address.
+    /// @param _deputySignature Deputy signature.
+    function setDeputy(address _deputy, bytes memory _deputySignature) external {
+        // Can only be called by the Foundation Safe itself.
+        if (msg.sender != address(FOUNDATION_SAFE)) {
+            revert DeputyPauseModule_NotFromSafe();
+        }
+
+        // Set the deputy address.
+        _setDeputy(_deputy, _deputySignature);
+    }
+
+    /// @notice Sets the DeputyGuardianModule.
+    /// @param _deputyGuardianModule DeputyGuardianModule address.
+    function setDeputyGuardianModule(IDeputyGuardianModule _deputyGuardianModule) external {
+        if (msg.sender != address(FOUNDATION_SAFE)) {
+            revert DeputyPauseModule_NotFromSafe();
+        }
+        deputyGuardianModule = _deputyGuardianModule;
+        emit DeputyGuardianModuleSet(_deputyGuardianModule);
+    }
+
+    /// @notice Calls the Foundation Safe's `execTransactionFromModuleReturnData()` function with
+    ///         the arguments necessary to call `pause()` on the Security Council Safe, which will
+    ///         then cause the Security Council Safe to trigger SuperchainConfig pause.
+    ///         Front-running this function is completely safe, it'll pause either way.
+    /// @param _nonce Signature nonce.
+    /// @param _signature ECDSA signature.
+    function pause(bytes32 _nonce, bytes memory _signature) external {
+        // Make sure the nonce hasn't been used yet.
+        if (usedNonces[_nonce]) {
+            revert DeputyPauseModule_NonceAlreadyUsed();
+        }
+
+        // Verify the signature.
+        bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(PAUSE_MESSAGE_TYPEHASH, PauseMessage(_nonce))));
+        if (ECDSA.recover(digest, _signature) != deputy) {
+            revert DeputyPauseModule_Unauthorized();
+        }
+
+        // Mark the nonce as used.
+        usedNonces[_nonce] = true;
```

Comment:
Should we emit an event to more easily let us figure out which nonces have been used? I don't have a concrete use case here, just generally good hygiene to emit events for storage changes

Reply:
Yeah that seems fine